commit 3e65551d0ab4acafaf4012601987b3391b2411ed
Author: Kefu Chai <kchai@redhat.com>
Date:   Sun Aug 25 14:53:20 2019 +0800

    mgr/ServiceMap: use plain gid
    
    * partially revert 5c25a018. which is not backward compatible.
    * change `ServiceMap::get_daemon()` so it returns a
      `pair<Daemon*,bool>`.
    
    git is an `optional<uint64_t>`, so we cannot dump it without checking.
    
    Fixes: https://tracker.ceph.com/issues/41424
    Signed-off-by: Kefu Chai <kchai@redhat.com>

diff --git a/src/mgr/ServiceMap.cc b/src/mgr/ServiceMap.cc
index 7870d222786..2a29e0ed900 100644
--- a/src/mgr/ServiceMap.cc
+++ b/src/mgr/ServiceMap.cc
@@ -40,7 +40,7 @@ void ServiceMap::Daemon::dump(Formatter *f) const
 {
   f->dump_unsigned("start_epoch", start_epoch);
   f->dump_stream("start_stamp") << start_stamp;
-  f->dump_unsigned("gid", *gid);
+  f->dump_unsigned("gid", gid);
   f->dump_string("addr", addr.get_legacy_str());
   f->open_object_section("metadata");
   for (auto& p : metadata) {
diff --git a/src/mgr/ServiceMap.h b/src/mgr/ServiceMap.h
index 7ca9fbd8688..3b6dd72068c 100644
--- a/src/mgr/ServiceMap.h
+++ b/src/mgr/ServiceMap.h
@@ -12,15 +12,13 @@
 #include "include/buffer.h"
 #include "msg/msg_types.h"
 
-#include <boost/optional.hpp>
-
 namespace ceph {
   class Formatter;
 }
 
 struct ServiceMap {
   struct Daemon {
-    boost::optional<uint64_t> gid;
+    uint64_t gid = 0;
     entity_addr_t addr;
     epoch_t start_epoch = 0;   ///< epoch first registered
     utime_t start_stamp;       ///< timestamp daemon started/registered
@@ -117,9 +115,11 @@ struct ServiceMap {
   void dump(ceph::Formatter *f) const;
   static void generate_test_instances(std::list<ServiceMap*>& ls);
 
-  Daemon* get_daemon(const std::string& service,
-		     const std::string& daemon) {
-    return &services[service].daemons[daemon];
+  std::pair<Daemon*,bool> get_daemon(const std::string& service,
+				     const std::string& daemon) {
+    auto& s = services[service];
+    auto [d, added] = s.daemons.try_emplace(daemon);
+    return {&d->second, added};
   }
 
   bool rm_daemon(const std::string& service,
